### PR TITLE
[WP-C1] add flashloan initiator check

### PR DESCRIPTION
**Pull request by dxganta**

### CODE REVIEW

File: src/sc4626.sol [hunk 1]
```diff
@@ -57,4 +65,16 @@ abstract contract sc4626 is ERC4626, AccessControl {
...
+        flashLoanInitiated = true;
+    }
+
+    function _finalizeFlashLoan() internal {
+        flashLoanInitiated = false;
+    }
+
+    function _isFlashLoanInitiated() internal view {
+        if (!flashLoanInitiated) revert InvalidFlashLoanCaller();
+    }
```
Comment (coolhill) [inline src/sc4626.sol:79]: sc4626 is for all types of strategies, not only those who use flashloans imo

File: src/steth/scUSDC.sol [hunk 2]
```diff
@@ -215,7 +215,9 @@ contract scUSDC is sc4626, IFlashLoanRecipient {
         uint256[] memory amounts = new uint256[](1);
         amounts[0] = debt - wethBalance;
 
+        _initiateFlashLoan();
         balancerVault.flashLoan(address(this), tokens, amounts, abi.encode(collateral, debt));
+        _finalizeFlashLoan();
```
Comment (coolhill) [inline src/steth/scUSDC.sol:220]:
```suggestion
        flashLoanInitiated = true;
        balancerVault.flashLoan(address(this), tokens, amounts, abi.encode(collateral, debt));
        flashLoanInitiated = false;
```
cheaper

Comment (dxganta) [reply src/steth/scUSDC.sol:220]: how much cheaper though?

Comment (coolhill) [reply src/steth/scUSDC.sol:220]: <table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Function Name</th>
<th scope="col" class="org-right">min</th>
<th scope="col" class="org-right">avg</th>
<th scope="col" class="org-right">median</th>
<th scope="col" class="org-right">max</th>
<th scope="col" class="org-right"># calls</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">scUSDC receiveFlashLoan</td>
<td class="org-right">2349</td>
<td class="org-right">93607</td>
<td class="org-right">94143</td>
<td class="org-right">183793</td>
<td class="org-right">4</td>
</tr>


<tr>
<td class="org-left">without function call</td>
<td class="org-right">2349</td>
<td class="org-right">93592</td>
<td class="org-right">94119</td>
<td class="org-right">183783</td>
<td class="org-right">4</td>
</tr>


<tr>
<td class="org-left">scWETH receiveFlashLoan</td>
<td class="org-right">2219</td>
<td class="org-right">292628</td>
<td class="org-right">225998</td>
<td class="org-right">488233</td>
<td class="org-right">54</td>
</tr>


<tr>
<td class="org-left">without function call</td>
<td class="org-right">2219</td>
<td class="org-right">293499</td>
<td class="org-right">226020</td>
<td class="org-right">488220</td>
<td class="org-right">54</td>
</tr>
</tbody>
</table>

its negligible i suppose we can keep it as is

File: src/sc4626.sol [hunk 3]
```diff
@@ -57,4 +65,16 @@ abstract contract sc4626 is ERC4626, AccessControl {
...
+
+    function _initiateFlashLoan() internal {
+        flashLoanInitiated = true;
+    }
+
+    function _finalizeFlashLoan() internal {
+        flashLoanInitiated = false;
+    }
+
+    function _isFlashLoanInitiated() internal view {
```
Comment (coolhill) [inline src/sc4626.sol:77]: this function can still revert altho you declare it `view`?

Comment (coolhill) [reply src/sc4626.sol:77]: seems unintuitive but i guess this is how it works, since you have the unit test

Comment (dxganta) [reply src/sc4626.sol:77]: yeah, the compiler suggested it to me.

**Review by brkomir — APPROVED** [review]

**Review by coolhill — APPROVED** [review]

**Pull request merged**